hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -645,16 +640,16 @@ def start_example(self, i: int, label_index: int) -> None:
 
     class _mutator_groups(ExampleProperty):
         def begin(self) -> None:
-            self.groups: "Dict[Tuple[int, int], List[int]]" = defaultdict(list)
+            self.groups: "Dict[Tuple[int, int], List[int]]" = defaultdict(set)
 
         def start_example(self, i: int, label_index: int) -> None:
-            depth = len(self.example_stack)
-            self.groups[label_index, depth].append(i)
+            key = (self.examples[i].ir_start, self.examples[i].ir_end)
+            self.groups[label_index].add(key)
 
         def finish(self) -> Iterable[Iterable[int]]:
             # Discard groups with only one example, since the mutator can't
             # do anything useful with them.
-            return [g for g in self.groups.values() if len(g) >= 2]
+            return [list(g) for g in self.groups.values() if len(g) >= 2]
```

Comment:
Nondeterministic iteration order is fine; I added the list cast to support `random.choice`. Fair point though that `set` is the ideal api to expose here instead of `list`, and we can just use `random.choice(list(groups))` at usage sites.

Reply:
I'd rather sort; these lists are short and someone might want a random seed to mean their process is reproducible.

Reply:
forgot about set ordering determinism, let's do sorted then.